File: php/bin/pqc_check.php
<?php
// Verifies the PHP port's post-quantum interop against the shared vector
// (spec/corpus/transcripts/pqc-interop.json), produced by the Java reference
// (BouncyCastle).
//
// It verifies the ML-DSA-65 signature over the vector's message using the
// vector's public key, exercising the node's real Crypto::mldsa65Verify (the
// openssl 3.5 CLI). Success proves PHP and Java agree on ML-DSA-65 byte-for-byte.
//
// It does NOT decapsulate the vector's ML-KEM ciphertext: the vector ships a
// 2400-byte FIPS *expanded* decapsulation key, while this port (like Go and JS)
// works from openssl's seed-keyed private form. That is a private-key
// *representation* difference, not an interop gap — a decapsulation key never
// crosses a node (only the encapsulation key, ciphertext, and public artifacts
// do, all standard FIPS encodings). Live ML-KEM-768 interop between PHP and the
// other implementations is proven directly by the interop matrix.
require __DIR__ . '/../src/autoload.php';

use Bonemesh\Crypto;

$path = $argv[1] ?? null;
if ($path === null) {
    fwrite(STDERR, "usage: pqc_check <pqc-interop.json>\n");
    exit(2);
}
$v = json_decode(file_get_contents($path), true)['mldsa65'];

$ok = Crypto::mldsa65Verify(hex2bin($v['public_hex']), hex2bin($v['message_hex']), hex2bin($v['signature_hex']));
if (!$ok) {
    fwrite(STDERR, "FAIL: PHP did not verify the Java ML-DSA-65 signature\n");
    exit(1);
}
echo "PASS: PHP verifies the Java ML-DSA-65 signature over the shared vector\n";
echo "NOTE: ML-KEM-768 interop with Java is proven live by the interop matrix\n";
echo "      (openssl's ML-KEM private form is seed-keyed; the vector's expanded\n";
echo "      dk is a key-representation detail that never crosses a node).\n";

File: rust/src/bin/pqc_check.rs
//! Cross-language post-quantum interop check: verifies the Java-produced vector
//! (spec/corpus/transcripts/pqc-interop.json) from the Rust side, using the
//! RustCrypto ml-dsa and ml-kem crates. Success means Rust verifies a signature
//! Java made and decapsulates a ciphertext Java produced to the identical
//! secret — post-quantum interop across implementations.

use std::process::exit;

use bonemesh::crypto;
use serde_json::Value;

fn hex_decode(s: &str) -> Vec<u8> {
    (0..s.len()).step_by(2).map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap()).collect()
}

fn main() {
    let path = std::env::args().nth(1).expect("path to pqc-interop.json");
    let doc: Value = serde_json::from_str(&std::fs::read_to_string(&path).unwrap()).unwrap();

    let dsa = &doc["mldsa65"];
    let dsa_ok = crypto::mldsa65_verify(
        &hex_decode(dsa["public_hex"].as_str().unwrap()),
        &hex_decode(dsa["message_hex"].as_str().unwrap()),
        &hex_decode(dsa["signature_hex"].as_str().unwrap()),
    );
    println!("ML-DSA-65: Rust verifies Java signature: {dsa_ok}");

    let kem = &doc["mlkem768"];
    let ss = crypto::mlkem_decapsulate(
        &hex_decode(kem["decapsulation_key_hex"].as_str().unwrap()),
        &hex_decode(kem["ciphertext_hex"].as_str().unwrap()),
    );
    let kem_ok = ss == hex_decode(kem["shared_secret_hex"].as_str().unwrap());
    println!("ML-KEM-768: Rust decapsulates Java ciphertext to Java secret: {kem_ok}");

    if dsa_ok && kem_ok {
        println!("post-quantum interop confirmed (Java -> Rust)");
    } else {
        eprintln!("post-quantum interop FAILED");
        exit(1);
    }
}
